### ISSUE TITLE
Trouble with typeclass reslution

### ISSUE THREAD

**Issue by RalfJung** [math-comp/math-comp#80]
We have a case of a diverging `apply:` related to typeclass resolution. We are trying to apply a lemma of following type:
```
∀ (M : ucmraT) (P Q : uPred M), PersistentP Q → (P ⊢ Q) → P ⊢ P ∗ Q
```
where `PersistentP` is a typeclass. We apply (`apply: lemma`) this to an evar-free goal of the right shape `own γ a ⊢ own γ a ∗ ✓ a`. This diverges. The output of `Set Typeclasses Debug` shows that it is looking for an instance of `PersistentP ?u`:
```
<prompt>own_valid_r < 89 |own_valid_r| 0 < </prompt>Set Typeclasses Debug.

<prompt>own_valid_r < 91 |own_valid_r| 0 < </prompt>apply: uPred.always_entails_r.
<infomsg>Debug: 1.1: apply @big_sepS_empty_persistent on (PersistentP ?u0)
</infomsg>
```
This is with ssrelfect 23e57fb47874331c5feaace883513b7abecdff28 and Coq 3a51aa7265f35dd3cbf3f7bff858d663e4406146.

Given the nature of the typeclass (deriving a property of assertions), it is not surprising that this query diverges -- coming up with any random assertion that is persistent makes no sense. However, I feel there is a already a bug here: We applied the lemma to a goal which determines `P` and `Q`. There is no reason to search for `PersistentP ?u`, it should be searching for `PersistentP (✓ a)` instead, which will easily succeed.

The same error happens when using ssreflect 1.6 and Coq 8.5. If we go back to an older version of our development, the search worked:
```
Debug: 1.1: apply @big_sepS_empty_persistent on (PersistentP ?u0)
Debug: 1.1.1.1: apply @fin_dec on (EqDecision ?Goal0)
Debug: 1.1.2.1: apply @finite.finite_countable on (Countable (fin ?Goal1))
Debug: 1.1.2.1.1.1: apply finite.fin_finite on (finite.Finite (fin ?Goal0))
Debug: 1.1: apply @cmra_valid_persistent on (PersistentP (✓ a))
```
I don't know why the first search terminates here, but the fact that it is even starting this search is already wrong.

To reproduce this yourself, the offending place in our development is <https://gitlab.mpi-sws.org/FP/iris-coq/blob/a9f7431c844f79b7df201529de75893b1e0d057e/base_logic/lib/own.v#L77>.
* Clone `https://gitlab.mpi-sws.org/FP/iris-coq.git`
* Go to commit a9f7431c844f79b7df201529de75893b1e0d057e
* Edit line 77 `base_logic/lib/own.v` to replace `apply (uPred.always_entails_r _ _)` by `apply: uPred.always_entails_r`
* See that statement diverge

**Comment by gares** [comment]
I think the "bug" is in the missing "mode" directive for your class.  You could try `Hint Mode PersistentP +.`

In the past I've tried to avoid this kind of problem, even in `apply:`, but I would not be surprised if there was yet another piece of code that triggers TC resolution even if I'm explicitly turning it off.  The Hint Mode directive should fix your bug once and forall.  It was there in 8.5 too, but not documented, so you should be able to use it.  FTR, it has nothing to do with SSR.

Does it solve your issue?

**Comment by gares** [comment]
Oh, and the fact that it used to terminate is probably because the order/precedence of TC instances changed.  Anyway, each and every TC declaration should be followed by a mode, in my humble opinion, or the generative semantics of "SLD resolution" gets in the way.

**Comment by RalfJung** [comment]
> I think the "bug" is in the missing "mode" directive for your class. You could try Hint Mode PersistentP +.

Yes, that works around the issue successfully. I should have mentioned that, sorry. It breaks other parts of the development (e.g. when we are searching for `PersistentP (inv ?x ?y)`). However, that can be fixed by using `!` instead of `+`, which has a saner semantics (introduced in Coq 8.6).

However, I see this as a work-around. We are not asking for `PersistentP ?x` to be resolved, so ssreflect shouldn't put the burden of not making that diverge on us. Notice that `apply (uPred.always_entails_r _ _)` works fine, it does not trigger this bogus typeclass search but only the one we want.

**Comment by gares** [comment]
I need time (that I don't have today) to see exactly who triggers TC resolution in this case, as I said before there are already some counter measures in `apply:`.

**Comment by RalfJung** [comment]
I just made an interesting observation, maybe it is helpful: If I change the shape of the goal to be `(P ⊢ Q) → P ⊢ P ∗ Q`, then `apply:` with the lemma described in the first post only triggers the expected typeclass search. If however I introduce the first assumption to turn the goal into just `P ⊢ P ∗ Q`, then `apply:` with the same lemma shows the behavior described above.

**Comment by gares** [comment]
`apply: F` is mostly doing `refine F || refine (F _) || refine (F _ _) || ...` .
If you are in debugging mood, try adding `Set SsrDebug.` to your script.
You should see the attempt calls to `refine`.  Which one loops?

**Comment by RalfJung** [comment]
`refine (F _ _)` loops. This is the output of `SsrDebug`:
```
<prompt>own_valid_r < 104 |own_valid_r| 0 < </prompt>Set SsrDebug.

<prompt>own_valid_r < 106 |own_valid_r| 0 < </prompt>apply: uPred.always_entails_r.
[[[ Diverging Typeclass Search Happens Here ]]]
1 focused subgoal
(shelved: 1), subgoal 1 (ID 28335)
  
  Σ : gFunctors
  A : cmraT
  H : inG Σ A
  γ : gname
  a : A
  ============================
  own γ a ⊢ ✓ a

(dependent evars:)

SSR: c@interp_refine=(always_entails_r (own γ a) (✓ a)
                        ?X28334@{__:=Σ; __:=A; __:=H; __:=γ; __:=a})
SSR: ==PF_ABS_EVARS_PIRREL==
SSR: c0= (always_entails_r (own γ a) (✓ a)
            ?X28334@{__:=Σ; __:=A; __:=H; __:=γ; __:=a})
SSR: evlist=?X28334
SSR: c0= (always_entails_r (own γ a) (✓ a)
            ?X28334@{__:=Σ; __:=A; __:=H; __:=γ; __:=a})
SSR: res= (λ _evar_0_ : own γ a ⊢ ✓ a,
           always_entails_r (own γ a) (✓ a) _evar_0_)
SSR: after: (λ _evar_0_ : own γ a ⊢ ✓ a,
             always_entails_r (own γ a) (✓ a) _evar_0_)

<prompt>own_valid_r < 107 |own_valid_r| 0 < </prompt>
```
I marked the spot where the (diverging) typeclass search happens.

After cancelling (C-c C-c in ProofGeneral), it seems ssreflect actually cancels only the current `refine` attempt and goes on with the next one. I reported this against Coq as https://coq.inria.fr/bugs/show_bug.cgi?id=5204, but by now I think this is actually a bug in `apply:`.

**Comment by CohenCyril** [comment]
@gares and @RalfJung has this issue been transferred to Coq?

**Comment by RalfJung** [comment]
Not that I know of.
